# Project 1 Lab notebook

 _To run code directly from notebook one should place it in project directory with `/data` subfolder in it._

##### 1. Download reference genome in .fasta and .gff formats in `/data` folder.

https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz
https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz

##### 2. Download shotgun sequence data in `/data` folder.

https://figshare.com/ndownloader/files/23769689

https://figshare.com/ndownloader/files/23769692

###### 3. Check data manually

In [26]:
! zcat data/amp_res_1.fastq.gz | head

@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101
GGTTGCAGATTCGCAGTGTCGCTGTTCCAGCGCATCACATCTTTGATGTTCACGCCGTGGCGTTTAGCAATGCTTGAAAGCGAATCGCCTTTGCCCACACG
+
@?:=:;DBFADH;CAECEE@@E:FFHGAE4?C?DE<BFGEC>?>FHE4BFFIIFHIBABEECA83;>>@>@CCCDC9@@CC08<@?@BB@9:CC#######
@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101
GTATGAGGTTTTGCTGCATTCTCTGNGCGAATATTAACTCCNTNNNNNTTATAGTTCAAAGCAAGTACCTGTCTCTTATACACATCTCCGAGCCCACGAGC
+
@@<?=D?D==?<AFGDF+AIHEACH#22<:?E8??:9??GG#0#####000;CF=C)4.==CA@@@)=7?C7?E37;3@>;;(.;>AB#############
@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101
GCTTCTCTTAACTGAGGTCACCATCATGCCGTTAAGTCCCTACCTCTCTTTTGCCGGTAACTGTTCCGCCGCGATTGCCTTTTATCTGTCTCTTATACACC

gzip: stdout: Broken pipe


In [27]:
! zcat data/amp_res_2.fastq.gz | head

@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101
GATCTAAGCTGAAGCCAGGCCAAAGTTTGACGATTGGTGCAGGCAGTAGCGCACAGCGACTGGCAAACAACAGCGATAGCATTACGTATCGTGTGCGCAAA
+
???BDB:DFHBFD@9;;+A;AFGH;ABHFHHGE@9:B:??@D>@;F?D8<<F8AA9EHHD8'..;5?A?A992(',(59CC3@C>22::A238+2>B<>B<
@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101
GTACTTGCTTTGNACTATAATATGCACGGAGNTAATATTCGCTCAGAGAATGCAGCAAAACCTCATACCTGTCTCTTATACACATCTGACGCTGCCGACGA
+
;@@DB?B;CFBB#2<:CB:FH<C@:<A?C::#1:86:BG9:8?8688?888EBF;783)=6-7=CC;ECD);?7;;>>AE;>(5;->AC@;B@;8?#####
@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101
ATAATAGGCAATCGCGTCGGAACAGTTACCGGCCAAAGAGAGGCAGGGACTTAACGGCATGATGGTGACCTCAGTTAAGAGAAGCCTGTCTCTTATACACA

gzip: stdout: Broken pipe


In [28]:
! zcat data/GCF_000005845.2_ASM584v2_genomic.fna.gz | head

>NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTG
GTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGAC
AGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGT
AACGGTGCGGGCTGACGCGTACAGGAAACACAGAAAAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTTCGACCAAAGG
TAACGAGGTAACAACCATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGTGTTGCCG
ATATTCTGGAAAGCAATGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCACCTGGTG
GCGATGATTGAAAAAACCATTAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGTATTTTTGCCGAACTTTT
GACGGGACTCGCCGCCGCCCAGCCGGGGTTCCCGCTGGCGCAATTGAAAACTTTCGTCGATCAGGAATTTGCCCAAATAA
AACATGTCCTGCATGGCATTAGTTTGTTGGGGCAGTGCCCGGATAGCATCAACGCTGCGCTGATTTGCCGTGGCGAGAAA

gzip: stdout: Broken pipe


Unarchive sequencing data.

In [ ]:
! gunzip -k data/amp_res_1.fastq.gz
! gunzip -k data/amp_res_2.fastq.gz

gzip: data/amp_res_1.fastq already exists; do you wish to overwrite (y or n)? ^C
gzip: data/amp_res_2.fastq already exists; do you wish to overwrite (y or n)? 

Estimated number of reads in each file are:

In [ ]:
! expr $(wc -l data/amp_res_2.fastq | cut -d " " -f 1) / 4 

##### 4. Install and run fastqc with sequence data; check output

`sudo apt-get install fastqc`

In [1]:
! fastqc -o . data/amp_res_1.fastq data/amp_res_2.fastq

Started analysis of amp_res_1.fastq
Approx 5% complete for amp_res_1.fastq
Approx 10% complete for amp_res_1.fastq
Approx 15% complete for amp_res_1.fastq
Approx 20% complete for amp_res_1.fastq
Approx 25% complete for amp_res_1.fastq
Approx 30% complete for amp_res_1.fastq
Approx 35% complete for amp_res_1.fastq
Approx 40% complete for amp_res_1.fastq
Approx 45% complete for amp_res_1.fastq
Approx 50% complete for amp_res_1.fastq
Approx 55% complete for amp_res_1.fastq
Approx 60% complete for amp_res_1.fastq
Approx 65% complete for amp_res_1.fastq
Approx 70% complete for amp_res_1.fastq
Approx 75% complete for amp_res_1.fastq
Approx 80% complete for amp_res_1.fastq
Approx 85% complete for amp_res_1.fastq
Approx 90% complete for amp_res_1.fastq
Approx 95% complete for amp_res_1.fastq
Analysis complete for amp_res_1.fastq
Started analysis of amp_res_2.fastq
Approx 5% complete for amp_res_2.fastq
Approx 10% complete for amp_res_2.fastq
Approx 15% complete for amp_res_2.fastq
Approx 20% c

Check out generated _.html_ files. 

Estimated read count is the same as calculated by fastqc (455876). "Per base sequence quality" is red circled in both files and "Per tile sequence quality" is red circled in forward sequence file (`amp_res_1`).

It may be useful to filter out low quality bases. What to do with the whole tiles that provide low quality data is unclear.

`amp_res_1_fastqc.html`

`amp_res_2_fastqc.html`



###### 5. Filter low quality reads with Trimmomatic

Install Trimmomatic with `sudo apt-get install trimmomatic`

Run it with following parameters:

* Delete adapters - `ILLUMINACLIP:/usr/share/trimmomatic/TruSeq3-PE.fa:2:30:10`

* Cut bases off the start of a read if quality below 20 - `LEADING:20`

* Cut bases off the end of a read if quality below 20 - `TRAILING: 20`

* Trim reads using a sliding window approach, with window size 10 and average quality  within the window 20 - `SLIDINGWINDOW:10:20`

* Drop the read if it is below length 20 - `MINLEN:20`

*not shure if adapter sequences are set correct; just copypasted it from http://www.usadellab.org/cms/?page=trimmomatic


In [ ]:
! TrimmomaticPE -phred33 data/amp_res_1.fastq data/amp_res_2.fastq \
data/amp_res_1_trimmed.fastq \
data/forward_unpaired.fastq \
data/amp_res_2_trimmed.fastq \
data/reverse_unpaired.fastq \
ILLUMINACLIP:/usr/share/trimmomatic/TruSeq3-PE.fa:2:30:10 \
LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20 

Check how many reads in trimmed files:

In [ ]:
! expr $(wc -l data/amp_res_1_trimmed.fastq | cut -d " " -f 1) / 4 

In [ ]:
! expr $(wc -l data/amp_res_2_trimmed.fastq | cut -d " " -f 1) / 4 

Generate fastqc reports

In [2]:
! fastqc -o . data/amp_res_1_trimmed.fastq data/amp_res_2_trimmed.fastq

Started analysis of amp_res_1_trimmed.fastq
Approx 5% complete for amp_res_1_trimmed.fastq
Approx 10% complete for amp_res_1_trimmed.fastq
Approx 15% complete for amp_res_1_trimmed.fastq
Approx 20% complete for amp_res_1_trimmed.fastq
Approx 25% complete for amp_res_1_trimmed.fastq
Approx 30% complete for amp_res_1_trimmed.fastq
Approx 35% complete for amp_res_1_trimmed.fastq
Approx 40% complete for amp_res_1_trimmed.fastq
Approx 45% complete for amp_res_1_trimmed.fastq
Approx 50% complete for amp_res_1_trimmed.fastq
Approx 55% complete for amp_res_1_trimmed.fastq
Approx 60% complete for amp_res_1_trimmed.fastq
Approx 65% complete for amp_res_1_trimmed.fastq
Approx 70% complete for amp_res_1_trimmed.fastq
Approx 75% complete for amp_res_1_trimmed.fastq
Approx 80% complete for amp_res_1_trimmed.fastq
Approx 85% complete for amp_res_1_trimmed.fastq
Approx 90% complete for amp_res_1_trimmed.fastq
Approx 95% complete for amp_res_1_trimmed.fastq
Analysis complete for amp_res_1_trimmed.fastq

Check .html files. Defect tile is still present, but sequence quality is now ok.

`amp_res_1_trimmed_fastqc.html`

`amp_res_2_trimmed_fastqc.html`


Let's try trim sequencing data with quality score 30:

In [ ]:
! TrimmomaticPE -phred33 data/amp_res_1.fastq data/amp_res_2.fastq \
data/amp_res_1_trimmed_30.fastq \
data/forward_unpaired_30.fastq \
data/amp_res_2_trimmed_30.fastq \
data/reverse_unpaired_30.fastq \
ILLUMINACLIP:/usr/share/trimmomatic/TruSeq3-PE.fa:2:30:10 \
LEADING:30 TRAILING:30 SLIDINGWINDOW:10:30 MINLEN:20 

In [ ]:
! expr $(wc -l data/amp_res_1_trimmed_30.fastq | cut -d " " -f 1) / 4 

In [ ]:
! expr $(wc -l data/amp_res_2_trimmed_30.fastq | cut -d " " -f 1) / 4 

In [25]:
! fastqc -o . data/amp_res_1_trimmed_30.fastq data/amp_res_2_trimmed_30.fastq

Started analysis of amp_res_1_trimmed_30.fastq
Approx 5% complete for amp_res_1_trimmed_30.fastq
Approx 10% complete for amp_res_1_trimmed_30.fastq
Approx 15% complete for amp_res_1_trimmed_30.fastq
Approx 20% complete for amp_res_1_trimmed_30.fastq
Approx 25% complete for amp_res_1_trimmed_30.fastq
Approx 30% complete for amp_res_1_trimmed_30.fastq
Approx 35% complete for amp_res_1_trimmed_30.fastq
Approx 40% complete for amp_res_1_trimmed_30.fastq
Approx 45% complete for amp_res_1_trimmed_30.fastq
Approx 50% complete for amp_res_1_trimmed_30.fastq
Approx 55% complete for amp_res_1_trimmed_30.fastq
Approx 60% complete for amp_res_1_trimmed_30.fastq
Approx 65% complete for amp_res_1_trimmed_30.fastq
Approx 70% complete for amp_res_1_trimmed_30.fastq
Approx 75% complete for amp_res_1_trimmed_30.fastq
Approx 80% complete for amp_res_1_trimmed_30.fastq
Approx 85% complete for amp_res_1_trimmed_30.fastq
Approx 90% complete for amp_res_1_trimmed_30.fastq
Approx 95% complete for amp_res_1_tr

`amp_res_1_trimmed_30_fastqc.html`

`amp_res_2_trimmed_30_fastqc.html`

It seems like not much has changed, but there is a lot more reads filtered, so it's probably not worth it to increase threshold. Note that in reverse sequence GC content is slightly changed, so yellow circle in FastQC report turned green.

##### 6. Align sequencing data on reference genome

Install BWA and Samtools:
    
`sudo apt-get install bwa`

`sudo apt-get install samtools`


Index reference genome.

In [ ]:
! bwa index data/GCF_000005845.2_ASM584v2_genomic.fna.gz

Align seqencing data on reference genome and compress it into .bam.

In [7]:
! bwa mem data/GCF_000005845.2_ASM584v2_genomic.fna.gz \
data/amp_res_1_trimmed.fastq \
data/amp_res_2_trimmed.fastq |\
samtools view -b \
> data/alignment.bam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 106302 sequences (10000070 bp)...
[M::process] read 108220 sequences (10000179 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 51224, 0, 21)
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (145, 185, 231)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 403)
[M::mem_pestat] mean and std.dev: (189.55, 63.63)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 489)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation RR...
[M::mem_pestat] (25, 50, 75) percentile: (117, 158, 1540)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4386)
[M::mem_pestat] mean and std.dev: (523.90, 696.25)
[M::mem_pestat] low and high boundaries for 

[M::mem_process_seqs] Processed 107254 reads in 5.692 CPU sec, 5.508 real sec
[M::process] read 37708 sequences (3545126 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (12, 50342, 0, 10)
[M::mem_pestat] analyzing insert size distribution for orientation FF...
[M::mem_pestat] (25, 50, 75) percentile: (68, 135, 213)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 503)
[M::mem_pestat] mean and std.dev: (120.09, 68.06)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 648)
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (144, 183, 229)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 399)
[M::mem_pestat] mean and std.dev: (188.31, 63.43)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 484)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation RR

Basic statistics:

In [12]:
! samtools flagstat data/alignment.bam

891306 + 0 in total (QC-passed reads + QC-failed reads)
891048 + 0 primary
0 + 0 secondary
258 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
890189 + 0 mapped (99.87% : N/A)
889931 + 0 primary mapped (99.87% : N/A)
891048 + 0 paired in sequencing
445524 + 0 read1
445524 + 0 read2
887108 + 0 properly paired (99.56% : N/A)
888960 + 0 with itself and mate mapped
971 + 0 singletons (0.11% : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


99.87% of reads are mapped.

Sort and index .bam file

In [18]:
! samtools sort data/alignment.bam > data/alignment_sorted.bam
! samtools index data/alignment_sorted.bam

##### 7. Variant calling

Create .mpileup file.

In [21]:
! samtools mpileup -f data/GCF_000005845.2_ASM584v2_genomic.fna data/alignment_sorted.bam >  data/my.mpileup

[mpileup] 1 samples in 1 input files


Write SNPs supported by at least *50%* of reads into .vcf file.

Nothing changes if set greater `--min-var-freq`. With 90% there are all same SNPs, but no indels, which are not shown in IGV anyway. 

In [1]:
! java -jar VarScan.v2.4.4.jar  \
mpileup2snp data/my.mpileup \
--min-var-freq 0.5 \
--variants --output-vcf 1 \
> data/VarScan_results.vcf

Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.2
Min avg qual:	15
P-value thresh:	0.01
Reading input from data/my.mpileup
4641514 bases in pileup file
9 variant positions (6 SNP, 3 indel)
0 were failed by the strand-filter
6 variant positions reported (6 SNP, 0 indel)


Manually inspected (IGV) mutated genes are:

acrB: https://www.uniprot.org/uniprotkb/P31224/entry : missense

ftsI: https://www.uniprot.org/uniprotkb/P0AD68/entry : missense 

rybA: https://www.ncbi.nlm.nih.gov/gene/2847681 : ?

mntP: https://www.uniprot.org/uniprotkb/Q8Z670/entry : missense

rsgA: https://www.uniprot.org/uniprotkb/P39286/entry : synonymous

envZ: https://www.uniprot.org/uniprotkb/P0AEJ4/entry : missense

##### 8. Automatic SNP annotation

Download [SnpEff](http://pcingola.github.io/SnpEff/) and extract `snpEff.jar`in project directory.

Download [`.gbff` file](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz) in `data/` subfolder.

Then make `.config` file:

In [12]:
! echo k12.genome : ecoli_K12 > snpEff.config 

Unarchive and rename reference genome file.

In [5]:
! mkdir data/k12
! gunzip data/GCF_000005845.2_ASM584v2_genomic.gbff.gz
! cp data/GCF_000005845.2_ASM584v2_genomic.gbff data/k12/genes.gbk 

Create database.

In [13]:
! java -jar snpEff.jar build -genbank -v k12

00:00:00 SnpEff version SnpEff 5.1d (build 2022-04-19 15:49), by Pablo Cingolani
00:00:00 Command: 'build'
00:00:00 Building database for 'k12'
00:00:00 Reading configuration file 'snpEff.config'. Genome: 'k12'
00:00:00 Reading config file: /home/quasicephalus/BI/Workshop/project1/snpEff.config
00:00:00 done
00:00:01 Chromosome: 'NC_000913.3'	length: 4641652
00:00:01 Create exons from CDS (if needed): 
..................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++..+++++++++.+++++++++++++++++++
	+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.
	+.++++..+++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++++

Annotate `.vcf` file and open it in IGV.

In [14]:
! java -jar snpEff.jar ann k12 data/VarScan_results.vcf > data/VarScan_results_annotated.vcf

SnpEff has confirmed manually obtained results, but now we have information about exact aminoacid changes in gene products:

ftsI : Ala544gly

acrB: Gln569Leu

mntP: Gly25Asp

envZ: Val241Gly

rsgA: Silent